src/validation.cpp

```diff
@@ -1159,6 +1159,20 @@ CTransactionRef GetTransaction(const CBlockIndex* const block_index, const CTxMe
 {
     LOCK(cs_main);
 
+    if (mempool && !block_index) {
+        CTransactionRef ptx = mempool->get(hash);
+        if (ptx) return ptx;
+    }
+    if (g_txindex) {
+        CTransactionRef tx;
+        uint256 block_hash;
+        if (g_txindex->FindTx(hash, block_hash, tx)) {
+            if (!block_index || block_index->GetBlockHash() == block_hash) {
+                hashBlock = block_hash;
+                return tx;
+            }
```

Comment:
```suggestion
            if (!block_index) {
                hashBlock = block_hash;
                return tx;
            }
            if (block_index->GetBlockHash() != block_hash) return nullptr;
```
This improves the performance of the case where a block hash is specified, but the transaction is found in a different block. The `if` at line 1170 would be false, so we would call `ReadBlockFromDisk()` below, but would not find the transaction there either (assuming the txindex is accurate).

I cherry-picked this PR with this suggestion onto #22437 as @jonatack [described above](https://github.com/bitcoin/bitcoin/pull/22383#issuecomment-879123985), and the tests all pass.